Add beginner, advanced, and contributors guide

canyon289 opened the pull request:
Putting up PR solely to facilitate discussion tomorrow. No intent to merge at the moment

Preview: https://jax--13285.org.readthedocs.build/en/13285/

canyon289 commented:
hi @jakevdp. At time of commenting theres three levels of reviews for this one, high level, per user profile level, and then technical details in that order. No immediate urgency, just prepping everything to make the most of our time on Friday

## Higher Level
Do the overall breakdown and general content of each of the seem amenable? I built them to the meeting we had, but given its been about two weeks just want to ensure I got it correct. 

If the user profiles seem correct then

## Lower Level
Each page will have its own content, but also tone and style. Within in each page I have questions as well, for instance
* In the Advanced Users page some tutorials are more "baked" than others. Should we keep or rearrange all of these?
* Should these tutorials then be removed from the front page if they are copied into the advanced users bucket
* Are there any that should be added, for instance some "hidden" notebooks that are in this repo but not in the docs

## Technical details and typo level
There are technical details to get htis work
* Getting responsive boxes added in `index.rst` to link to each of these
* Getting cross link working
* Fixing youtube element size etc
* Fix up any typos I left behind

I will fix these all after the upper two things are mostly decided.

8bitmp3 commented:
```
docs/advanced_guide.rst

...
Advanced Users
==============
```

Note that you're using "guides" and "tutorials" in the same file. Let's stick to "tutorials" and rename the RST file to avoid future confusion (see "common_gotchas" -> "Sharp Bits" doc).

I'd recommend removing "The following tutorials will provide advanced functionality in JAX" since it's kind of repeating what the title says. A list of advanced tutorials would do IMO.

Is `advanced_guide.rst` more like `index.rst` for _Advanced tutorials_?

8bitmp3 commented:
```
docs/beginner_guide.md
...

# Beginner Guide
Welcome to the beginners guide for JAX. 
On this page we will introduce you to the key ideas of Jax,
...
## Key Ideas of Jax
At a high level aims to empower researchers with a small, agile,
constantly evolving software system and excellent support.

Specifically Jax does this by providing researchers with

1. **Familiar API**: JAX provides a familiar NumPy-style API for ease of adoption by researchers and engineers.
2. **Transformations**: JAX includes composable function transformations for compilation, batching, automatic differentiation, and parallelization.
3. **Run Anywhere**: The same code executes on multiple backends, including CPU, GPU, & TPU

If you prefer a video introduction here's one from Jake Vanderplas...

## Installing JAX
Installation instructions ...
Alternatively Jax comes preinstalled on[Google Colab](https://colab.research.google.com/)...

## Getting familiar with Jax
For most users starting out the key functionalities of JAX to become familiar with are
* jit
* grad
* vmap

For these topics the (jax-101)[../jax-101/index.rst] and (Jax Quickstart)[../notebooks/quickstart]
provide code examples with greater explanation
```

This file may appear like a (second) landing page - an short introduction to JAX with high-level information and links to get started. This may introduce an extra layer of information "overload". I'm adding @cgarciae - he's helped with the Flax site. @marcvanzee (also from Flax) is proposing to reorganize the ToC on Flax too.

---

To solve the "we have 3 beginner guides" issue (cc @skye @jakevdp ) I can assist with combining the main ideas from 1) Quickstart, 2) JAX-101, and 3) Thinking in JAX and putting them in one document (we still keep the other docs to avoid 404s, especially the JAX-101 notebook 1, since it's part of the series of notebooks). Working on something similar for Flax (discussed with @marcvanzee).

8bitmp3 commented:
```
docs/contributing.rst

Contributing to JAX
===================

JAX welcomes contributions.
See below for various install guides to get setup as a developer
as well as how to submit Jax Enhancement Proposals for larger proposals.

.. toctree::
   :maxdepth: 1
   :caption: Developer documentation

   contributing
   developer
   jax_internal_api
   autodidax
   jep/index
```

For "Contributing", it may help external users start opening GitHub PRs/Issues if they can scan through all the instructions on one page. Check out Keras - https://keras.io/contributing/ - they have a lot of external examples submitted by the community via PRs. Also Numpy https://numpy.org/contribute/.

Each of the subsections - such as "Developer", "JAX API", "JEP" - would be on the same page. In Flax, this looks like https://flax.readthedocs.io/en/latest/advanced_topics/contributing.html (We 're moving this doc to a more visible location soon cc @@cgarciae  @marcvanzee )

jakevdp commented:
Great start - thanks for working on this!

8bitmp3 commented:
> Are there any that should be added, for instance some "hidden" notebooks that are in this repo but not in the docs

Hidden 💎 s. Good point. Let's discuss this (and everything else) this week.

canyon289 commented on docs/advanced_guide.rst:
```diff
@@ -0,0 +1,27 @@
+.. _advanced_guide:
+.. :orphan:
```
Regarding the warnings, getting file missing from toctree warnings that's supposed to be fixed by adding these orphan tags but it doesnt seem to be working. @8bitmp3 any idea how I can get around this?

<img width="1790" alt="image" src="https://user-images.githubusercontent.com/7213793/206567140-adfb627c-aac5-4b8d-a398-f049a8a58c6e.png">

canyon289 replied:
ill try it1 I knew there had to be some magic parameter!

jakevdp commented on docs/beginner_guide.rst:
```diff
@@ -0,0 +1,45 @@
+* grad
+* vmap
+
+For these topics the (jax-101)[../jax-101/index.rst] and (Jax Quickstart)[../notebooks/quickstart]
```
nitpick: let's highlight the quickstart first, then link to the JAX-101 tutorials for those who want a bit more depth.

canyon289 replied:
Sounds good!

jakevdp commented on docs/index.rst:
```diff
@@ -10,6 +10,17 @@ parallelize, Just-In-Time compile to GPU/TPU, and more.
+
+User Specific Guides
+********************
+
```
I think here is where you could use a hidden toctree to correctly index these guides

jakevdp commented on docs/index.rst:
```diff
@@ -46,32 +61,6 @@ parallelize, Just-In-Time compile to GPU/TPU, and more.
-
-   notebooks/autodiff_cookbook
-   multi_process
-   notebooks/Parallelism_with_JAX
```
It looks like this was removed here but not added in the other place. Maybe a sync issue, since this notebook was added relatively recently

canyon289 replied:
Ah so obvious. thanks for noting that

canyon289 commented:
@jakevdp @8bitmp3 This is ready for final review.

Regarding CI there seems to be some code checkout issue in the RTD CI runner. There's no errors locally though, so I hope therell be non in CI once it runs properly

<img width="1792" alt="image" src="https://user-images.githubusercontent.com/7213793/207388211-098289b7-d0ec-4390-bf5c-20a3d4bb0263.png">

jakevdp commented:
That issue should be fixed after #13636, if you sync to main and try again it should build I think

canyon289 commented:
Everything works 🎉

jakevdp commented on docs/index.rst:
```diff
@@ -10,6 +10,14 @@ parallelize, Just-In-Time compile to GPU/TPU, and more.
    guide for more information.
 
+
+User Specific Guides
```
Using a heading here makes the toctrees below this appear to be part of the heading.

Also, it's not clear how the "Beginner" link here is differentiated from "Getting Started" just below.

canyon289 replied:
Theres two levels of discussion here.

The first is whether we want to move these or not. The reason I had left them is because they were the most visited links off of the home page.

If we want to move them I propose Installing, Quickstart, and how to think go in the Beginner guide, and sharp bits go into advanced.

Happy to go either which way

jakevdp commented on docs/beginner_guide.rst:
```diff
@@ -0,0 +1,47 @@
+
+.. raw:: html
+
+	<iframe width="1776" height="840" src="https://www.youtube.com/embed/WdTeDXsOSj4" title="Intro to JAX: Accelerating Machine Learning research" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
```
Use a narrower width? 1800px is huge even on my large monitor

canyon289 replied:
Changed

jakevdp commented on docs/beginner_guide.rst:
```diff
@@ -0,0 +1,47 @@
+
+Installing JAX
+=================
+Installation instructions are available on the [Install Guide](https://github.com/google/jax#installation).
```
We need rst-style hyperlinks here. Alternatively, you could change this file to `beginner_guide.md` and then use myst-md syntax for the sphinx directives, whichever you prefer

canyon289 replied:
Fixed to RST style

jakevdp commented on docs/beginner_guide.rst:
```diff
@@ -0,0 +1,47 @@
+no installation required
+
+Getting familiar with Jax
+=============================
```
Sphinx sometimes complains if the size of underlines doesn't match the size of the text

canyon289 replied:
Fixed

jakevdp commented on docs/beginner_guide.rst:
```diff
@@ -0,0 +1,47 @@
+For most users starting out the key functionalities of JAX to become familiar with are
+* jit
+* grad
+* vmap
```
Use ```:func:`jax.jit` ``` to get cross-references here

canyon289 replied:
Done

jakevdp commented on docs/jax-101/index.rst:
```diff
@@ -1,3 +1,5 @@
+:orphan:
```
Still needed?

canyon289 replied:
Yes because its not in a toctree if I dont add it well get a warning

canyon289 commented:
Thanks @jakevdp for finding the Jupyter cards. This brings us really close to the envisioned design. I'm not the best visual designer so I'm sure there's ways we can make these cards look much better, I may be able to ask one of the colleagues on my team or ask someone else internally if they wouldn't mind some light consulting

@8bitmp3 Please do any wordsmithing or title changes for the categories. They're now referenced in a couple of places (filenames, titles, css classes) so it'll be great to align and make the changes once!

Appreciate the support from you both


<img width="891" alt="image" src="https://user-images.githubusercontent.com/7213793/208252266-013613d6-7cbb-4460-b8d3-86ae5c3b9c50.png">

Reference documentation
https://sphinx-design.readthedocs.io/en/latest/cards.html#cards-options
https://m3.material.io/components/cards/

jakevdp commented on docs/_static/style.css:
```diff
@@ -3,3 +3,15 @@
+
+.contributor-guide {
+  background-color: #ab00b6;
+}
```
I'd probably lighten these background colors to make it easier to read the text (perhaps use RGBA with an opacity of 50%?)

canyon289 replied:
Great idea! Added

jakevdp commented on docs/contributor_guide.rst:
```diff
@@ -0,0 +1,21 @@
+
+JAX welcomes contributions from the community.
+See below for various install guides to get setup as a developer
+as well as how to submit Jax Enhancement Proposals for larger proposals.
```
```suggestion
as well as developer-focused resources such as Jax Enhancement Proposals.
```

jakevdp commented on docs/index.rst:
```diff
@@ -10,6 +10,45 @@ parallelize, Just-In-Time compile to GPU/TPU, and more.
+
+         Beginner Guide
+         ^^^
+         New to Jax. Start here
```
Wordsmithing: 
```
Getting Started
^^^
New to JAX? Start here
```

jakevdp commented on docs/index.rst:
```diff
@@ -10,6 +10,45 @@ parallelize, Just-In-Time compile to GPU/TPU, and more.
+
+         Advanced Users
+         ^^^
+         Dive deeper 
```
Wordsmithing
```
Advanced Usage
^^^
Dive deeper into JAX
```

jakevdp commented on docs/index.rst:
```diff
@@ -10,6 +10,45 @@ parallelize, Just-In-Time compile to GPU/TPU, and more.
+
+         Contributing 
+         ^^^
+         How to contribute to JAX
```
Wordsmithing
```
Developer Resources
^^^
Contributing to JAX
```

canyon289 commented:
All comments addressed. Thanks Jake!

<img width="953" alt="image" src="https://user-images.githubusercontent.com/7213793/208547855-5de11054-7cb3-4f7e-93b9-db297878d573.png">

canyon289 commented:
Anything else I can do to keep this moving?

8bitmp3 commented:
To recap the discussion with @jakevdp and @canyon289 :

- We can move the three key points from https://jax--13285.org.readthedocs.build/en/13285/beginner_guide.html#beginner-guide to the main index.rst after the "Note: JAX 0.4.1  introduces new parallelism APIs..." banner.
- Buttons on the main index.rst: "Getting started", "Guides" (instead of "Advanced..."), and "Developer resources" (what is "advanced" is hard to explain, similar to Flax).
- Keep sentences in buttons short (without the second sentences).
- The ToC in the main index.rst is the current user journey:

```
Getting started
- Installing JAX
- JAX Quickstart
- How to think in JAX
- JAX - The Sharp Bits
- Tutorial: JAX 101
- Runtime value debugging
```

so once more content is added above this in the main index.rst, some users may not be able to view that "learning path".

8bitmp3 commented:
Thank you so much @canyon289 :+1:

canyon289 commented:
Made the changes though. It now feels like the top section is fairly busy between a paragraph, announcement, and the three additional boxes.

For hte other points
* Deleted the sentence
* If there is a better word to use than advanced feel free to make the suggested edit in the rst file. I dont have any better ideas
* For the last one if thats the case not much more should be added over the user journey. Then itd stay on screen

<img width="989" alt="image" src="https://user-images.githubusercontent.com/7213793/211114999-7978377e-a64e-4903-b7d2-ceb4882a3265.png">

canyon289 commented:
I'm looking over the flax vs numpy docs I'm realizing whats going on.

The original idea was to add a guide of "where to go" similar to Numpy does, and move links to appropriate areas.
The Flax docs, and these new additions, are now spending more front page real estate telling people "what flax is", and now in this case Jax.

Theres also now this third thing of the note being added as well, warning users of a design change.

Neither idea is wrong, but right now its feeling like this is the subpar combination of the two, and the temporary warning is further muddling things.

https://flax.readthedocs.io/en/latest/
https://numpy.org/doc/stable/

jakevdp commented on docs/index.rst:
```diff
@@ -10,6 +10,66 @@ parallelize, Just-In-Time compile to GPU/TPU, and more.
+
+   .. card:: 
+
+      Familiar API
```
I think you can do something like
```
.. card:: Familiar API

   JAX provides a familiar NumPy-style API for ease of adoption by researchers and engineers. 
```
to make the title recognized as a title

canyon289 replied:
Tried that. thought the horizontal line looked nice as a separator, though the bolding is nice too. Here they are side by side.

I don't have a strong preference. Do you like one of the two better? Happy to change it

<img width="826" alt="image" src="https://user-images.githubusercontent.com/7213793/211118231-1801c615-c6ad-496f-94e1-4aa6ffdc6cbe.png">

jakevdp replied:
We're into aesthetic bikeshedding now, but my preference would be to make them all like the one on the right, and with the borders removed if possible

canyon289 replied:
Nothing a bit of css cant fix! Hows this look? And thank you for the rapid feedback, well converge much faster!

<img width="833" alt="image" src="https://user-images.githubusercontent.com/7213793/211119420-3a5cc921-3cde-453a-8db3-545bf27a92c5.png">

jakevdp replied:
Looks great to my eyes 😁

canyon289 replied:
🎉 Pushed a latest set of commits to fix the build error. Want to be sure we get this done by monday for that meeting. 

Thanks again for rapid reviews

jakevdp commented:
Hey Ravin - it looks like somehow the "Key Ideas" grid is now oriented vertically. Can you take a look?

canyon289 commented:
Thanks for catching that

canyon289 commented:
Looks like a case "But it works on my computer..." 

Ill figure it out today and ping this when verified correct in Ci.

jakevdp commented:
Looks great - mind if I share with the dev team for comments this morning?

canyon289 commented:
> Looks great - mind if I share with the dev team for comments this morning?

Please do so! Happy to keep moving this forward so we can ensure everyone is happy and this keeps moving forward

canyon289 commented on docs/beginner_guide.rst:
```diff
@@ -0,0 +1,46 @@
+
+If looking to jump straight in take a look at the jax quickstart.
+
+.. toctree::
```
I couldnt figure out how to create a direct reference to a .ipynb file from sphinx so settling for this toc tree instead for now

jakevdp replied:
Yeah, we would need myst-md anchors to reference notebook sections (e.g. `(reference-name)=`) but unfortunately these are not hidden in the notebook view. So elsewhere we tend to just refer to URLs, which mostly never go stale.

Using a TOC is fine here.

canyon289 replied:
Thanks

canyon289 commented on docs/beginner_guide.rst:
```diff
@@ -0,0 +1,46 @@
+- :func:`jax.grad` 
+- :func:`jax.vmap` 
+
+For introductions to specific topics take a look at the tutorials contained in Jax-101
```
Moved up Jax 101 docs to have higher prominence

jakevdp commented:
I think this is looking pretty good - would you mind squashing and rebasing? I'm leaning toward merging and then iterating more from there.

jakevdp reviewed (approved):
Thanks for all the work on this!

I'm thinking we should merge it and then continue iterating.

jakevdp commented:
Looks like readthedocs hit a transitory HTTP error accessing intersphinx resources.

canyon289 commented:
> Looks like readthedocs hit a transitory HTTP error accessing intersphinx resources.

Anything I can do to fix it? Doesnt seem like I can restart the job from my end, or at least I can't see how

jakevdp commented:
I've been poking around this, and I have a couple ideas. First, I think the main page could still be cleaned up a bit. I've been playing with some ideas in this branch https://github.com/jakevdp/jax/commit/8411308d6d5743d68a28d40d076bd61a43859a09

In particular, I made the toctrees on the main page hidden, so that the content displays in the left sidebar. I think the fact that the advanced guide material no longer shows up in the sidebar is problematic. We probably don't need it listed in full, but perhaps it could be under a heading that is collapsed by default?

Regarding the new beginner link, I think it needs some work before we can make it live – after all our edits, it's no longer very well organized, and I don't think linking to `jit`/`vmap`/`grad` API docs is useful. Instead, we should focus on using it to guide people to the existing beginner docs

Let me know what you think - thanks!

canyon289 commented:
> I've been poking around this, and I have a couple ideas. First, I think the main page could still be cleaned up a bit. I've been playing with some ideas in this branch [jakevdp@a4c4544](https://github.com/jakevdp/jax/commit/a4c4544ffc2c21974549065c97c8e09269663d61)
> 
> In particular, I made the toctrees on the main page hidden, so that the content displays in the left sidebar, but not at the bottom of the index page. I think the fact that the advanced guide material no longer shows up in the sidebar is problematic. We probably don't need it listed in full, but perhaps it could be under a heading that is collapsed by default?
> 
> Regarding the new beginner link, I think it needs some work before we can make it live – after all our recent edits, it's no longer very well organized, and I don't think linking prominently to `jit`/`vmap`/`grad` API docs is useful to a new user. Instead, we should focus on using it to guide people to the existing beginner docs as appropriate.
> 
> Let me know what you think - thanks!

Thanks for these direct edits!
1. I'm thrilled to see the front page simplified even more, with the links deduplicated. This was going to be my next suggestion, originally didn't want to edit to the main page too much to avoid controversy but hey lets go for it.
2. In the beginner page having a bunch of taking up for jit, vmap, grad isn't useful. I do think calling out that those three things are the core ideas to start with though is useful. If its a concern of space would putting them into a sentence with inline links suffice, or at a minimum just stating it in plaintext?


<img width="1290" alt="image" src="https://user-images.githubusercontent.com/7213793/211702893-3a918bb4-ccea-4d08-9962-aa1e2f3bc62d.png">

jakevdp commented:
I think linking to them is fine (maybe in a paragraph?) My main concern was that the page doesn't offer a very good entrypoint for beginners at the moment – I called those out as examples of why this is the case

canyon289 commented:
> I think linking to them is fine (maybe in a paragraph?) My main concern was that the page doesn't offer a very good entrypoint for beginners at the moment – I called those out as examples of why this is the case

Makes sense, Removed the links. In latest set of commits I also
* Expanded main page TOC tree on the left
* Simplified getting started page some more

I didnt have any strong idea of how to order the TOC tree on the left so its more random than thought out right now. First wanted to check if it was comprehensive first.

canyon289 commented:
> Looks awesome. Ping me when you are at the finish line for the last review @canyon289

Thanks! In that case will wait for an LGTM from Jake (or any final edits) and then we can do the final pass to get this in. Appreciate it

jakevdp commented:
Sorry, I'm going to take a look at this but I've been delayed by other things

jakevdp commented:
I suggested some modifications in https://github.com/jakevdp/jax/commit/dbf24d2dbc38f07020f8d916c54e756b7380fc6f The key difference is that now all content appears on the index sidebar, though most is collapsed by default.

jakevdp commented on docs/index.rst:
```diff
@@ -1,105 +1,104 @@
 
-   debugging/index
+      .. card:: :material-regular:`library_books;2em` User Guides
+         :link: user-guide
```
This link needs to be `user-guides` now I think

jakevdp replied:
Looks like the change didn't make it into the commit?

canyon289 replied:
Youre right my apologies

jakevdp replied:
Looks like you updated the `class-card` but not the `link`

canyon289 committed:
Update doc landing page

Co-authored-by: 8bitmp3 <19637339+8bitmp3@users.noreply.github.com>
Co-authored-by: Jake VanderPlas <jakevdp@google.com>

jakevdp commented:
Thanks for all the work on this!
